lib/spack/spack/provider_index.py

```diff
@@ -2,54 +2,134 @@
+        # Add all the providers that satisfy the vpkg spec.
+        if virtual_spec.name in self.providers:
+            for p_spec, spec_set in self.providers[virtual_spec.name].items():
+                if p_spec.satisfies(virtual_spec, deps=False):
+                    result.update(spec_set)
 
-       { 'mpi' :
-           { mpi@:1.1 : set([mpich]),
-             mpi@:2.3 : set([mpich2@1.9:]) } }
+        # Return providers in order. Defensively copy.
+        return sorted(s.copy() for s in result)
 
-       Calling providers_for(spec) will find specs that provide a
-       matching implementation of MPI.
+    def __contains__(self, name):
+        return name in self.providers
 
-    """
+    def satisfies(self, other):
+        """Return True if the providers of virtual specs are compatible,
+        False otherwise.
 
-    def __init__(self, specs=None, restrict=False):
-        """Create a new ProviderIndex.
+        Args:
+            other: another provider index
+        """
+        common = set(self.providers) & set(other.providers)
+        if not common:
+            return True
+
+        # This ensures that some provider in other COULD satisfy the
+        # vpkg constraints on self.
+        result = {}
+        for name in common:
+            crossed = _cross_provider_maps(
+                self.providers[name], other.providers[name]
+            )
+            if crossed:
+                result[name] = crossed
 
-        Optional arguments:
+        return all(c in result for c in common)
 
-        specs
-            List (or sequence) of specs.  If provided, will call
-            `update` on this ProviderIndex with each spec in the list.
+    def __eq__(self, other):
+        return self.providers == other.providers
 
-        restrict
-            "restricts" values to the verbatim input specs; do not
-            pre-apply package's constraints.
+    def _transform(self, transform_fun, out_mapping_type=dict):
+        return _transform(self.providers, transform_fun, out_mapping_type)
 
-            TODO: rename this.  It is intended to keep things as broad
-            as possible without overly restricting results, so it is
-            not the best name.
+    def __str__(self):
+        return str(self.providers)
+
+    def __repr__(self):
+        return repr(self.providers)
+
+
+class ProviderIndex(_IndexBase):
+    def __init__(self, specs=None, restrict=False):
+        """Provider index based on a single mapping of providers.
+
+        Args:
+            specs (list of specs): if provided, will call update on each
+                single spec to initialize this provider index.
+
+            restrict: "restricts" values to the verbatim input specs; do not
+                pre-apply package's constraints.
+
+        TODO: rename this.  It is intended to keep things as broad
+        TODO: as possible without overly restricting results, so it is
+        TODO: not the best name.
```

Comment:
Did you intend to come back and rename this here or in the other PR or leave as-is since it was from a previous commit?

Reply:
I would leave that as is for now, unless somebody has a good suggestion for that :slightly_smiling_face: